### ISSUE TITLE
Allow grouping of commands

### ISSUE THREAD

**Issue by Loilo** [yargs/yargs#684]
It's amazing that we have the possibility to group options under a certain headline.
It would be great though to be able to do the same with commands.

In case anyone's not convinced how that would be useful, I'll explain my personal use case:

I made a CLI tool that has some built-in commands but is pluggable in a way so commands can be added from outside.
I'd like to kind of set apart those commands from the built-in ones in the `--help`, maybe under a "Plugins" headline.

The most convenient way in my eyes would be to be able to group commands.

What do you think?

**Comment by nexdrew** [comment]
I like this idea. Unfortunately I think the commands API is a little cumbersome when it comes to adding new features (the latest at the time of this writing was [_command aliases_](http://yargs.js.org/docs/index.html#methods-commandmodule-command-aliases)).

For command modules, it's no problem, but for non-module commands, we have to either get creative or come up with a better API (something other than just tacking on new method arguments).

I'll give this some thought.

**Comment by wmnnd** [comment]
Would you accept a pull request that adds `commandGroup(opts, groupName)` that would work like `group(opts, groupName)` but for commands?

**Comment by itai-codefresh** [comment]
Any updates about this?
It is a very needed feature.
@wmnnd will you be able to expose your code so maybe we can use it?

**Comment by wmnnd** [comment]
@itai-codefresh Nope, I ended up using `nopt` instead. You can check out [dblsqd-cli](https://www.npmjs.com/package/dblsqd-cli) to see how I implemented it.

**Comment by quilicicf** [comment]
That feature would be awesome indeed. I tried looking at the code to see if I could come up with an implementation but there are things I don't quite understand in the code.

From what I gather, a method `commandGroup` should be added in `yargs.js`. That method should work like `group` does but for commands, with a `getCommandGroups` method to get them. 

Where I'm getting lost is with the `preservedGroups` and `freeze` stuff. I don't see tests related to that part of the code which makes it a bit harsh to understand. 

I could probably try to implement something but I guess I'd need a little helping hand.

**Comment by bcoe** [comment]
I continue to think this would be a slick feature, if anyone wants to take it on.

**Comment by osd-henry** [comment]
+1 on this. I recently built an extensible CLI app with growing number of commands. The command groups (with an alternative `--help`'s command list generation) will make make things more comprehensible.

**Comment by hungluu** [comment]
+1 on this. I recently built an extensible CLI app with growing number of commands. The command groups (with an alternative / or customisable `--help`'s command list generation) will make make things more comprehensible.

**Comment by chrfrenning** [comment]
+1 - still relevant 2ish years later - anything that grows beyond 4-5 commands will benefit from this in terms of usability.

**Comment by jcs090218** [comment]
+1

**Comment by stackblaze** [comment]
I'm not sure why this issue was closed without addressing people's requests.

**Comment by shadowspawn** [comment]
@stackblaze it has not been closed (that was a linked issue)

**Comment by markus-incisive** [comment]
+1 still very much wanted. Currently hijacking the `help` command, and parsing its default output to achieve this.

```typescript
// Add commands
const parser = yargs()
  .command(/* ... */)
  .command(/* ... */)

// Create a `help` command, then grab the value of `getHelp()` that the `help` command will use
const help = await parser
  .help(false)
  .command({
    command: 'help',
    describe: 'Show help',
    handler: () => {
      console.log(help.replace(/Commands:\n( {2}.+\n)*/g, match => {
        // Grab all the commands in the "Commands:" section of the help output
        const { commands } = match.split('\n').slice(1).reduce((accumulator, line) => {
          if (/^ {2}\S/.test(line)) {
            if (accumulator.current.length > 0) {
              const help = accumulator.current.join('')
              accumulator.commands.push({ command: help.match(/^ {2}(\S+)/)?.[1] ?? '', help })
            }
            accumulator.current = [line + '\n']
          } else {
            accumulator.current.push(line + '\n')
          }
          return accumulator
        }, {
          current: [] as string[],
          commands: [] as Array<{ command: string, help: string }>
        })

        // Sort the commands into groups however you like
        const groups: Record<string, Array<{ order: number, help: string }>> = {}
        const getGroup = (commandName: string): { group: string, index: number } | null => {
          /* ... */
        }
        for (const { command, help } of commands) {
          const { group, index } = getGroup(command) ?? { group: 'Other', index: 0 }
          ;(groups[group] ?? (groups[group] = [])).push({ order: index, help })
        }

        // Create the string that replaces the "Commands:" section
        return Object.entries(groups).map(([group, commands]) =>
          `${group}:\n` + commands.sort((a, b) => a.order - b.order).map(({ help }) => help).join('')
        ).join('\n')
      }))
    }
  })
  .getHelp()

// Execute
await parser.parseAsync(process.argv.slice(2))
```
